library/ssl_cli.c

```diff
@@ -2398,24 +2398,18 @@ static int ssl_parse_server_ecdh_params_psa( mbedtls_ssl_context *ssl,
         return( MBEDTLS_ERR_SSL_ILLEGAL_PARAMETER );
     handshake->ecdh_bits = (uint16_t) ecdh_bits;
 
-    /*
-     * Put peer's ECDH public key in the format understood by PSA.
-     */
-
+    /* Keep a copy of the peer's public key */
     ecpoint_len = *(*p)++;
     if( (size_t)( end - *p ) < ecpoint_len )
         return( MBEDTLS_ERR_SSL_DECODE_ERROR );
 
-    if( mbedtls_psa_tls_ecpoint_to_psa_ec(
-                                    *p, ecpoint_len,
-                                    handshake->ecdh_psa_peerkey,
-                                    sizeof( handshake->ecdh_psa_peerkey ),
-                                    &handshake->ecdh_psa_peerkey_len ) != 0 )
-    {
-        return( MBEDTLS_ERR_SSL_HW_ACCEL_FAILED );
-    }
+    if( ecpoint_len > sizeof( handshake->ecdh_psa_peerkey ) )
+        return( MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE );
```

Comment:
Hmm, I think both make sense, but IMO `HANDSHAKE_FAILURE` is slightly better. The most likely explanation for why this test would fail is the the server sent an invalid message. We know our buffer is large enough to hold a pubkey for all the curves that are enabled in the build, and we know that the curve chosen by the server is on that list, so if the `ecpoint_len` sent by the server is larger that our buffer, it means the server's message is inconsistent, as the `ecpoint_len` doesn't match the chosen curve.

So, IMO that falls under "the server's parameters don't make sense", so "The handshake negotiation failed." (the description of `HANDSHAKE_FAILURE`). I agree that `MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL` "A buffer is too small to receive or write a message" also makes sense but in that case I was more inclined to point out that it's the server's fault, and enlarging the buffer won't actually help, as the server's message is still going to be invalid, so the handshake would fail even with a larger buffer - only it would fail later, when we try to actually use the public key (rather than just store a copy): `psa_raw_key_agreement()` is going to complain that `ecdh_psa_peerkey_len` doesn't match the chosen curve.